### PR TITLE
feat: improve preload and suspense integration

### PR DESCRIPTION
**Pull request by promer94**
This pr improves the integration between `preload` and suspense mode. 

`preload` primarily used to solve the waterfall problem of suspense, but now it also could be used to implement retry for suspense.

close: #160 
close: #602 
close: #1907 
close: #1093 
close: #2239

### Implementation
1. allowing `useSWR` to directly access the cached `PRELOAD` promise when suspense. 
2. call `mutate` with the `PRELOAD` promise to update the data cache.

### [Example](https://codesandbox.io/p/sandbox/suspense-retry-ss4m7k)
```tsx
import { Suspense } from 'react'
import { ErrorBoundary } from 'react-error-boundary'
import useSWR, { preload } from 'swr'

let count = 0
export const fetcher = () => {
  count++
  if (count === 1) return Promise.reject('wrong')
  return Promise.resolve({ name: 'success' })
}

const key = 'manual-retry'

const Demo = () => {
  const { data } = useSWR(key, fetcher, {
    suspense: true
  })
  return <div>data: {data?.name}</div>
}

function Fallback({ resetErrorBoundary }) {
  return (
    <div role="alert">
      <p>Something went wrong</p>
      <button
        onClick={resetErrorBoundary}
      >
        retry
      </button>
    </div>
  )
}

function App() {
  return (
    <div className="App">
      <ErrorBoundary FallbackComponent={Fallback} onReset={() => preload(key, fetcher)}>
        <Suspense fallback={<div>loading</div>}>
          <Demo />
        </Suspense>
      </ErrorBoundary>
    </div>
  )
}
```

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -17,6 +17,8 @@
     "target": "ES2018",
     "baseUrl": ".",
     "noEmitOnError": true,
+    /** FIXME: only use this to skip build error */
+    "downlevelIteration": true,
```
Comment (huozhi) [inline tsconfig.json:21]: we didn't enable this by purpose https://github.com/vercel/swr/pull/2181

File: _internal/utils/mutate.ts [hunk 2]
```diff
@@ -73,8 +74,7 @@ export async function internalMutate<Data>(
     const keyFilter = _key
     const matchedKeys: Key[] = []
     const it = cache.keys()
-    for (let keyIt = it.next(); !keyIt.done; keyIt = it.next()) {
-      const key = keyIt.value
+    for (const key of it) {
```
Comment (huozhi) [inline _internal/utils/mutate.ts:77]: Needs to revert due to #2181

Comment (promer94) [reply _internal/utils/mutate.ts:77]: we have already upgraded the build target to es2018 and we also already use for of in https://github.com/vercel/swr/blob/7badbb9d428d2ac48873b3a3f11d54e7cec0c7cf/_internal/utils/cache.ts#L68

I also have checked the build result and it seemed like the for of part was kept.

I guess it won't be a problem here ?

Comment (huozhi) [reply _internal/utils/mutate.ts:77]: should be safe now with es2018. then i think you can just remove the FIXME comment in tsconfig

**Review by huozhi — APPROVED** [review]

**Pull request merged**